### PR TITLE
implement parallel training

### PR DESCRIPTION
**Pull request by zenoone**
This pull request implements parallel training on multiple-gpus. Operations peformend on the batch like computing the average energy or resampling walkers can be synchronized over gpus. To reduce communication each process creates its own log files. If the processes are synced tensorboard writers are combined.

### CODE REVIEW

File: src/deepqmc/fit.py [hunk 1]
```diff
@@ -182,31 +187,35 @@ def fit_wf(  # noqa: C901
...
-            E_loc_loss_mean, E_loc_loss_var = weighted_mean_var(Es_loc_loss, log_ws)
+            E_loc_loss_mean, E_loc_loss_var = batch_weighted_mean_var(
+                Es_loc_loss, log_ws
+            )
             writer.add_scalar('E_loc_loss/mean', E_loc_loss_mean, step)
             writer.add_scalar('E_loc_loss/var', E_loc_loss_var, step)
             writer.add_scalar('loss', loss, step)
-            writer.add_scalar('log_weights/KLvar', log_ws.var() / 2, step)
+            log_ws_mean, log_ws_var = batch_weighted_mean_var(log_ws)
+            writer.add_scalar('log_weights/mean', log_ws_mean, step)
```
Comment (jhrmnn) [inline src/deepqmc/fit.py:209]: this value doesn't really have any information value

File: src/deepqmc/torchext/ddp.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from torch.nn.parallel import DistributedDataParallel
```
Comment (jhrmnn) [inline src/deepqmc/torchext/ddp.py:1]: let's put this in `deepqmc.torchext.utils`

File: src/deepqmc/cli.py [hunk 3]
```diff
@@ -246,3 +251,79 @@ def evaluate_at(workdir, cuda, store_steps, hook):
...
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12345'
+    dist.init_process_group('nccl', rank=rank, world_size=world_size)
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def _train_at_parallel(rank, workdir, save_every, n_gpus, state, sync_processes):
```
Comment (jhrmnn) [inline src/deepqmc/cli.py:266]: is it essentially `train_at()` plus setting the `PARALLEL` flag minus the restart logic?

**Review by jhrmnn — CHANGES_REQUESTED** [review]

File: src/deepqmc/cli.py (hunk 3 above)
Comment (zenoone) [reply src/deepqmc/cli.py:266]: Its essentially `train_at()` with the PARALLEL flag and some additional setup for running the training on multiple gpus. This involes the `setup()` and the `cleanup()` functions as well as `mp.spawn`. For now it doesn't contain the restart logic, but I will add that in eventually. Also, this version of the cli will be deprecated and replaced with the hydrabased cli, so this is meant as a temporary solution and will be updated. I will furthermore add an `evaluate_parallel`  function, that facilitates running an evaluation on multiple gpus of a device.

Comment (jhrmnn) [reply src/deepqmc/cli.py:266]: I was just wondering if some of the code could be just straight shared with `train_at()`. But that's overthinking it at this point.

**Comment by jhrmnn** [comment]
Looks good now. Can you look into how easy would it be to add some tests of the parallel functionality? If it turns out too involved, then don't bother for now.

File: src/deepqmc/torchext/batch_operations.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+def batch_exp_normalize_mean(t):
+    t_max = batch_max(t)
+    t_shifted = t - t_max
+    return t_shifted.exp() / batch_mean(t_shifted.exp())
+
+
+def batch_weighted_mean_var(t, log_ws=None):
+    log_ws = torch.zeros_like(t) if log_ws is None else log_ws
+    ws = batch_exp_normalize_mean(log_ws)
```
Comment (kmaziarz) [inline src/deepqmc/torchext/batch_operations.py:55]: Technically we could skip this if `log_ws` was `None`, and set `log_ws` to `torch.ones_like` directly, but it doesn't really matter I guess

File: src/deepqmc/sampling.py [hunk 5]
```diff
@@ -334,21 +347,21 @@ def step(self):
...
             self.writer.add_scalar('sampling/acceptance', acceptance, self._step_writer)
             self.writer.add_scalar('sampling/tau', self.tau, self._step_writer)
             self.writer.add_scalar(
-                'sampling/age/max', info['age'].max(), self._step_writer
+                'sampling/age/max', batch_max(info['age']).item(), self._step_writer
             )
             self.writer.add_scalar(
                 'sampling/age/rms',
-                np.sqrt((info['age'] ** 2).mean()),
+                batch_mean(torch.sqrt(info['age'] ** 2)).item(),
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/sampling.py:364]: Good catch, thx.

File: src/deepqmc/torchext/batch_operations.py (hunk 4 above)
Comment (zenoone) [reply src/deepqmc/torchext/batch_operations.py:55]: Sounds reasonable, if have changed it accordingly.

File: src/deepqmc/fit.py [hunk 6]
```diff
@@ -182,31 +187,34 @@ def fit_wf(  # noqa: C901
...
+        E_loc_err = torch.sqrt(E_loc_var / batch_len(Es_loc))
         lr = opt.state_dict()['param_groups'][0]['lr']
         if writer:
             writer.add_scalar('E_loc/mean', E_loc_mean, step)
             writer.add_scalar('E_loc/var', E_loc_var, step)
-            writer.add_scalar('E_loc/min', Es_loc.min(), step)
-            writer.add_scalar('E_loc/max', Es_loc.max(), step)
+            writer.add_scalar('E_loc/min', batch_min(Es_loc), step)
+            writer.add_scalar('E_loc/max', batch_max(Es_loc), step)
             writer.add_scalar('E_loc/err', E_loc_err, step)
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/fit.py:201]: Yes, all processes are doing the logging. This does not take longer than logging only from a single process, because all other processes have to wait anyways, and reduces the necessairy code changes.

File: src/deepqmc/fit.py [hunk 7]
```diff
@@ -158,7 +163,7 @@ def fit_wf(  # noqa: C901
                 log.warning('Masking local energies where psi = 0')
                 Es_loc = Es_loc.where(mask, Es_loc.new_tensor(0))
             Es_loc_loss = log_clipped_outliers(Es_loc, q) if clip_outliers else Es_loc
-            loss = loss_func(Es_loc_loss, log_psis, exp_normalize_mean(log_ws))
+            loss = loss_func(Es_loc_loss, log_psis, batch_exp_normalize_mean(log_ws))
             # The convention is that `loss_func` returns an *average* loss over
             # all the inputs. We scale it so that it works with subbatching.
             loss *= len(rs) / batch_size
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/fit.py]: Yes you are right. In order to precisely resemble the single gpu training the loss should be scaled by the number of gpus.
